modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java

```diff
@@ -168,6 +185,35 @@ private Function<LeafReaderContext, CheckedIntFunction<List<Object>, IOException
                     "Field [" + name() + "] of type [" + CONTENT_TYPE + "] cannot run positional queries since [_source] is disabled."
                 );
             }
+            if (searchExecutionContext.isSourceSynthetic()) {
+                String name = originalFieldName();
+                return context -> docID -> {
+                    try {
+                        List<Object> values = new ArrayList<>();
+                        context.reader().document(docID, new StoredFieldVisitor() {
+                            private Status found = Status.NO;
+
+                            @Override
+                            public Status needsField(FieldInfo fieldInfo) {
+                                if (fieldInfo.name.equals(name)) {
+                                    found = Status.STOP;
+                                    return Status.YES;
+                                }
+                                return found;
+                            }
+
+                            @Override
+                            public void stringField(FieldInfo fieldInfo, String value) {
+                                assert fieldInfo.name.equals(name);
+                                values.add(value);
+                            }
+                        });
+                        return values;
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                };
```

Comment:
There's a stored field lookup thing in `searchExecutionContext.lookup()` but it can't be convinced to load the hidden stored field. If we feel strongly about it I can try and integrate into it, but I'm not super sure how at the moment.

Reply:
Yeah that's strictly for scripts and is integrated poorly with other stored field lookup stuff, so I don't think it's worth trying to re-use it for the moment.  I *do* think that the document lookup API I'm playing with at the moment will improve this though.

Reply:
searchExecutionContext.lookup()

Reply:
It has a lovely caching mechanism that I think could be quite nice. If multiple queries need to recheck the source it'll load once while this won't.

Reply:
So, somewhat annoyingly, I don't think we will be able to re-use a stored field loader from the SearchLookup here because the underlying API expects a LeafReaderContext, not a LeafSearchLookup.  But you should be able to use a LeafStoredFieldLoader rather than a FieldVisitor here which will at least be more readable.